app/Models/LoginAttempt.php

```diff
@@ -35,27 +36,136 @@ class LoginAttempt extends Model
 {
     protected $table = 'osu_login_attempts';
     protected $primaryKey = 'ip';
+    public $incrementing = false;
     public $timestamps = false;
 
+    public static function appendFailedIds($user, $state)
+    {
+        $userId = static::getUserId($user);
+        $newFailedId = DB::getPdo()->quote("{$userId}($state)");
+
+        return DB::raw("CONCAT(failed_ids, ',', {$newFailedId})");
+    }
+
+    public static function findOrDefault($ip)
+    {
+        try {
+            return static::find($ip) ?? static::create([
+                'ip' => $ip,
+                'failed_ids' => '',
+                'unique_ids' => 0,
+                'failed_attempts' => 0,
+                'total_attempts' => 0,
+            ]);
+        } catch (Exception $e) {
+            if (is_sql_unique_exception($e)) {
+                return static::findOrDefault($ip);
+            }
+        }
+    }
+
+    public static function getUserId($user)
+    {
+        return optional($user)->getKey() ?? 0;
+    }
+
+    public static function hashInvalidPassword($password)
+    {
+        // The goal is just to allow vaguely matching password - for
+        // example when trying same password - to be excluded from being
+        // counted as additional attempt.
+        return substr(sha1('osu_unique_'.md5($password)), 8, 12);
```

Comment:
But vaguely matching passwords would result in completely different hashes..?

Reply:
as in there might be another password resulting same "hash"

Reply:
Only same password and collisions would result in the same result, and the collisions won't even be from similar values since the password is being hashed? The chance of a vaguely matching password returning the same result might as well be non-existent

Reply:
it's not "vaguely matching password" but "vaguely matching" password (as in just because it matches doesn't mean it's the same) :dancer:

Reply:
it's not allowing anything then ಠ_ಠ